src/test/java/com/iota/iri/service/tipselection/impl/ReferenceCheckerImplTest.java

```diff
@@ -0,0 +1,146 @@
+        Assert.assertFalse(referenceChecker.doesReference(transactions.get(1).getHash(), transactions.get(0).getHash()));
+    }
+
+    @Test
+    public void testReferenceCheckerReturnsTrueForDirectApprovers() throws Exception {
+        List<TransactionViewModel> transactions = new ArrayList<>();
+
+        transactions.add(new TransactionViewModel(getRandomTransactionTrits(), getRandomTransactionHash()));
+        transactions.add(new TransactionViewModel(getRandomTransactionTrits(), getRandomTransactionHash()));
+        transactions.add(new TransactionViewModel(getRandomTransactionWithTrunkAndBranch(
+                transactions.get(0).getHash(),
+                transactions.get(1).getHash()),
+                getRandomTransactionHash()));
+
+        for (TransactionViewModel transaction : transactions) {
+            transaction.store(tangle);
+        }
+
+        ReferenceChecker referenceChecker = new ReferenceCheckerImpl(tangle);
+
+        Assert.assertTrue(referenceChecker.doesReference(transactions.get(2).getHash(), transactions.get(1).getHash()));
+        Assert.assertTrue(referenceChecker.doesReference(transactions.get(2).getHash(), transactions.get(0).getHash()));
+    }
+
+    @Test
+    public void testReferenceCheckerReturnsFalseForDirectApproversWhenReversingDirection() throws Exception {
+        List<TransactionViewModel> transactions = new ArrayList<>();
+
+        transactions.add(new TransactionViewModel(getRandomTransactionTrits(), getRandomTransactionHash()));
+        transactions.add(new TransactionViewModel(getRandomTransactionTrits(), getRandomTransactionHash()));
+        transactions.add(new TransactionViewModel(getRandomTransactionWithTrunkAndBranch(
+                transactions.get(0).getHash(),
+                transactions.get(1).getHash()),
+                getRandomTransactionHash()));
+
+        for (TransactionViewModel transaction : transactions) {
+            transaction.store(tangle);
+        }
+
+        ReferenceChecker referenceChecker = new ReferenceCheckerImpl(tangle);
+
+        Assert.assertFalse(referenceChecker.doesReference(transactions.get(0).getHash(), transactions.get(2).getHash()));
+        Assert.assertFalse(referenceChecker.doesReference(transactions.get(1).getHash(), transactions.get(2).getHash()));
+    }
+
+    @Test
+    public void testGetConfidencesThrowsExceptionWhenReferenceTransactionDoesNotExist() throws Exception {
+        List<TransactionViewModel> transactions = new ArrayList<>();
+        List<TransactionViewModel> nonExistentTransactions = new ArrayList<>();
+
+        transactions.add(new TransactionViewModel(getRandomTransactionTrits(), getRandomTransactionHash()));
+        transactions.add(new TransactionViewModel(getRandomTransactionTrits(), getRandomTransactionHash()));
+        nonExistentTransactions.add(new TransactionViewModel(getRandomTransactionTrits(), getRandomTransactionHash()));

```

Comment:
small nit - maybe move these below the store command (or even name them something w/ not stored) - that block at first glance looks like it will store all 4.